src/mame/snk/hng64_sprite.ipp

```diff
@@ -40,194 +40,424 @@
+	// Could also be draw order related, check if it inverts the z value?
+	bool zsort = !(m_spriteregs[0] & 0x01000000);
+
 	if (zsort)
-		std::stable_sort(m_spritelist.begin(), m_spritelist.end(), [] (auto const &a, auto const &b) { return a.first > b.first; });
+		m_sprite_zbuffer.fill(0x0000, cliprect);
+	else
+		m_sprite_zbuffer.fill(0x07ff, cliprect);
 
-	for(auto it : m_spritelist)
+	uint32_t* source = m_spriteram;
+	uint32_t* end = m_spriteram + (0xc000 / 4);
+
+	while (source < end)
 	{
-		source = it.second;
-
-		int tileno,chainx,chainy,xflip;
-		int pal,xinc,yinc,yflip;
-		uint16_t xpos, ypos;
-		int xdrw,ydrw;
-		int chaini;
-		uint32_t zoomx,zoomy;
-		float foomX, foomY;
-		int blend;
-
-		ypos = (source[0]&0xffff0000)>>16;
-		xpos = (source[0]&0x0000ffff)>>0;
+		uint16_t zval = (source[2] & 0x07ff0000) >> 16;
+
+		uint16_t ypos = (source[0] & 0xffff0000) >> 16;
+		uint16_t xpos = (source[0] & 0x0000ffff) >> 0;
 		xpos += (spriteoffsx);
 		ypos += (spriteoffsy);
 
-		tileno= (source[4]&0x0007ffff);
-		blend=  (source[4]&0x00800000);
-		yflip=  (source[4]&0x01000000)>>24;
-		xflip=  (source[4]&0x02000000)>>25;
-
-		pal =(source[3]&0x00ff0000)>>16;
+		int tileno = (source[4] & 0x0007ffff);
+		bool blend = (source[4] & 0x00800000);
+		bool checkerboard = (source[4] & 0x04000000);
+		int yflip = (source[4] & 0x01000000) >> 24;
+		int xflip = (source[4] & 0x02000000) >> 25;
 
-		chainy=(source[2]&0x0000000f);
-		chainx=(source[2]&0x000000f0)>>4;
-		chaini=(source[2]&0x00000100);
+		int pal = (source[3] & 0x00ff0000) >> 16;
 
-		zoomy = (source[1]&0xffff0000)>>16;
-		zoomx = (source[1]&0x0000ffff)>>0;
+		int chainy = (source[2] & 0x0000000f);
+		int chainx = (source[2] & 0x000000f0) >> 4;
+		int chaini = (source[2] & 0x00000100);
 
-#if 0
-		if (!(source[4] == 0x00000000 || source[4] == 0x000000aa))
-			osd_printf_debug("unknown : %.8x %.8x %.8x %.8x %.8x %.8x %.8x %.8x \n", source[0], source[1], source[2], source[3],
-				source[4], source[5], source[6], source[7]);
-#endif
+		uint32_t zoomy = (source[1] & 0xffff0000) >> 16;
+		uint32_t zoomx = (source[1] & 0x0000ffff) >> 0;
 
 		/* Calculate the zoom */
+		/* FIXME: regular zoom mode has precision bugs, can be easily seen in Samurai Shodown 64 intro */
+		int zoom_factor = (m_spriteregs[0] & 0x08000000) ? 0x1000 : 0x100;
+		if (!zoomx) zoomx = zoom_factor;
+		if (!zoomy) zoomy = zoom_factor;
+
+		s32 dx, dy;
+
+		if (zoom_factor == 0x100)
 		{
-			int zoom_factor;
+			dx = zoomx << 8;
+			dy = zoomy << 8;
+		}
+		else
+		{
+			dx = zoomx << 4;
+			dy = zoomy << 4;
+		}
 
-			/* FIXME: regular zoom mode has precision bugs, can be easily seen in Samurai Shodown 64 intro */
-			zoom_factor = (m_spriteregs[0] & 0x08000000) ? 0x1000 : 0x100;
-			if(!zoomx) zoomx=zoom_factor;
-			if(!zoomy) zoomy=zoom_factor;
+		uint32_t srcpix_x;
+		uint32_t srcpix_y;
```

Comment:
Please move the declaration of `srcpic_y` to where it’s used (immediately before the loop), and move the declaration of `srcpix_x` inside the loop as the value isn’t valid across loop iterations.